### PR TITLE
feat: add retrievetargetdir to source:retrieve

### PR DESCRIPTION
**Pull request by peternhale**
### What does this PR do?
add retrievetargetdir to source:retrieve
### What issues does this PR fix or reference?
@W-10735446@

### CODE REVIEW

File: src/promiseQueue.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*
```
Comment (mshanemc) [inline src/promiseQueue.ts:1]: overall comment...there's some tools that do this kind of thing if we find a lot more use cases for it.

https://github.com/supercharge/promise-pool is fairly small, zero-dep.  The error handling and abort looks nice.

File: src/commands/force/source/retrieve.ts [hunk 2]
```diff
@@ -220,4 +229,56 @@ export class Retrieve extends SourceCommand {
...
+    let overlapsProject = true;
+    const resolvedTargetDir = resolve(this.flags.retrievetargetdir as string);
+    overlapsProject = !!this.project.getPackageDirectories().find((pkgDir) => {
+      if (pkgDir.fullPath) {
+        return pkgDir.fullPath.includes(resolvedTargetDir);
+      }
+      return false;
+    });
+    if (overlapsProject) {
+      return;
```
Comment (mshanemc) [inline src/commands/force/source/retrieve.ts:271]: this feels like a good place for a warning.  Another option would be to make this "overlaps" validation a standalone function that could be used for your new flag's `parse` so that it doesn't go through all that retrieve work to fail this late in the game.

As is, it looks like the user would specify a targetDir and then if it overlapped, the retrieve would go to their default dir anyway without a warning?

File: src/commands/force/source/retrieve.ts [hunk 3]
```diff
@@ -220,4 +229,56 @@ export class Retrieve extends SourceCommand {
...
+        50
+      );
+      return directories;
+    }
+
+    if (!this.flags.retrievetargetdir) {
+      return;
+    }
+    let overlapsProject = true;
+    const resolvedTargetDir = resolve(this.flags.retrievetargetdir as string);
```
Comment (mshanemc) [inline src/commands/force/source/retrieve.ts:263]: and what if I want to "peek" at the results, outside my normal project.  So I say, `--retrievetargetdir peek/main/default`

what's going to happen?  `.replace(join('main', 'default')` would eliminate the redundancy?

**Review by mshanemc — CHANGES_REQUESTED** [review]
only firm request is readdir `withFileTypes`

everything else is comments, questions, and style preferences.

File: src/promiseQueue.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * @param sourceQueue - The list of items to process.
+ * @param producer - The function to produce a promise from an item.
+ * @param concurrency - The number of promises to run at a time.
+ * @param queueResults - Whether to queue the results of the promises.
+ */
+export async function promisesQueue<T>(
+  sourceQueue: T[],
+  producer: (T) => Promise<T | T[]>,
+  concurrency: number,
+  queueResults = false
```
[the remark this answers is not in the document]
Comment (peternhale) [reply src/promiseQueue.ts:19]: @mshanemc it is used in mv function. Yes it can and yes I have :). Not being able to queue results defeats the ability to accomplish the dir traversal without recursion.

File: src/commands/force/source/retrieve.ts (hunk 3 above)
Comment (peternhale) [reply src/commands/force/source/retrieve.ts:262]: I'll give that a try, but my suspect that it will only remove one `main/default`

File: src/commands/force/source/retrieve.ts (hunk 2 above)
Comment (peternhale) [reply src/commands/force/source/retrieve.ts:270]: @mshanemc I think I am going to change direction on overlapping targets, fail the command if the target overlaps a package dir. There are other commands and variants of source:retrieve to get source into a package dir.

File: src/commands/force/source/retrieve.ts (hunk 3 above)
Comment (peternhale) [reply src/commands/force/source/retrieve.ts:262]: As predicted, the results land in `peek/main/default`

**Review by mshanemc — APPROVED** [review]

File: messages/retrieve.json [hunk 5]
```diff
@@ -24,6 +26,11 @@
     "forceoverwrite": "ignore conflict warnings and overwrite changes to the project"
   },
   "flagsLong": {
+    "retrievetargetdir": [
+      "The root of the directory structure into which the source files are retrieved.",
+      "If the target directory matches one of the package directories in your sfdx-project.json file, the command will fail.",
```
Comment (jshackell-sfdc) [inline messages/retrieve.json:31]:
```suggestion
      "If the target directory matches one of the package directories in your sfdx-project.json file, the command fails.",
```

File: messages/retrieve.json [hunk 6]
```diff
@@ -24,6 +26,11 @@
...
+      "Running the command multiple times with the same target will add new files and overwrite existing files."
```
Comment (jshackell-sfdc) [inline messages/retrieve.json:32]:
```suggestion
      "Running the command multiple times with the same target adds new files and overwrites existing files."
```

File: messages/retrieve.json [hunk 7]
```diff
@@ -52,5 +59,6 @@
   "errorColumn": "PROBLEM",
   "nothingToRetrieve": "Specify a source path, manifest, metadata, or package names to retrieve.",
   "wantsToRetrieveCustomFields": "Because you're retrieving one or more CustomFields, we're also retrieving the CustomObject to which it's associated.",
-  "retrieveWontDelete": "You currently have files deleted in your org.  The retrieve command will NOT delete them from your local project"
+  "retrieveWontDelete": "You currently have files deleted in your org.  The retrieve command will NOT delete them from your local project",
+  "retrieveTargetDirOverlapsPackage": "The retrieve target directory [%s] overlaps one of your package directories.  Please specify a different retrieve target directory."
```
Comment (jshackell-sfdc) [inline messages/retrieve.json:63]:
```suggestion
  "retrieveTargetDirOverlapsPackage": "The retrieve target directory [%s] overlaps one of your package directories.  Specify a different retrieve target directory and try again."
```

**Comment by mshanemc** [comment]
QA notes:

✅  retrieve classes to new dir `out` 
✅ retrieve classes to a folder inside existing dir `out/main/default`
✅ retrieve classes to new folder with several levels inside existing dir `out2/main/default`
✅ retrieve over an existing file `--retrievetargetdir .gitignore`
`ERROR running force:source:retrieve:  Metadata API request failed: Component conversion failed: ENOTDIR: not a directory, mkdir '.gitignore/main'`
✅  retrieve non-existent type to new dir [doesn't leave empty dir]
✅  retrieve to existing non-project dir
✅  retrieve to overwrite pkgDir `--retrievetargetdir force-app`
`ERROR running force:source:retrieve:  The retrieve target directory [force-app] overlaps one of your package directories.  Specify a different retrieve target directory and try again.` (fails before mdapi call)

❌ retrieve to new child folder of pkgDir `--retrievetargetdir force-app/main/default/test`
I wish we blocked this, too.  Makes a mess of tracking (still seems to deploy, though) Also, it put the files in force-app/test (not where I said to put them) and created an empty folder where I expected the files to appear `force-app/main/default/test`

✅  retrieve to same dir twice

**Comment by mshanemc** [comment]
✅ force-app hits error
✅ force-ap works fine
✅ force-app/main hits error
✅ force-app/main/default/foo hits error
✅ force works fine

couldn't find any additional problems.  🎉
